DNN/ONNX: outputs registration regression, feature request for new version of Clip operator

##### System information (version)

- OpenCV => 4.5.5
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2019

##### Detailed description

I tried to infer using a Selfie Segmenter ONNX model (you can find the model here: https://github.com/PINTO0309/PINTO_model_zoo/tree/main/109_Selfie_Segmentation), however I get Nan on all output values.

##### Steps to reproduce

You can replicate this issue simply by running this simple script with OpenCV 4.5.5:

```
#include <iostream>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/core.hpp>

int main()
{
    cv::Size inputSizeNewBarracuda = cv::Size(256, 256);

    std::string imagefilename = "C:/Lixo/SantaNoel.jpg";
    std::string newBarracuda = "C:/Users/cesar.gouveia/Downloads/saved_model_openvino/model_float32.onnx";

    cv::dnn::Net net = cv::dnn::readNetFromONNX(newBarracuda);
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

    cv::Mat img = cv::imread(imagefilename);
    cv::Mat resized;
    cv::resize(img, resized, inputSizeNewBarracuda);

    std::vector<cv::Mat> imgBatch = { resized };
    bool swapRBChannels = false;
    cv::Mat blob = cv::dnn::blobFromImages(imgBatch, 1.0, cv::Size(), cv::Scalar(), swapRBChannels, false, CV_32F);
    blob = blob.reshape(1, { 1, inputSizeNewBarracuda.height, inputSizeNewBarracuda.width, 3 }); // because the model has input in channels last

    net.setInput(blob);

    std::vector<cv::Mat> outputs;
    outputs.clear();

    std::vector<cv::String> unconnectedOutLayerNames = net.getUnconnectedOutLayersNames();
    net.forward(outputs, unconnectedOutLayerNames);

    const cv::Mat& targetMat = outputs[0];
    const float* targetBuffer = (float*)targetMat.data;

    std::cout << targetMat.size[0] << std::endl; // 1
    std::cout << targetMat.size[1] << std::endl; // 256
    std::cout << targetMat.size[2] << std::endl; // 256
    std::cout << targetMat.size[3] << std::endl; // 1

    // Access contiguos buffer
    for (size_t i = 0; i < 256 * 256; i++)
    {
        std::cout << targetBuffer[i] << std::endl;
    }

    // Access Mat
    targetMat.reshape(1, { 1, 256, 256 });
    for (size_t o = 0; o < targetMat.size[0]; o++)
    {
        for (size_t i = 0; i < targetMat.size[1]; i++)
        {
            for (size_t j = 0; j < outputs[0].size[2]; j++)
            {
                std::cout << outputs[0].at<float>(o, i, j);
            }
        }
    }
    
    std::cout << "Finished" << std::endl;
}
```

The model can be downloaded by the link I provided. This is the first "channels last" model that I use with OpenCVDNN, all my other models are channel first and I never had this code behavior before. I tried to access the mat and the contiguos memory array but neither of them worked. 

Thanks,
César.

> "channels last" model

are you sure about this ? 
the [pdf here](https://github.com/PINTO0309/PINTO_model_zoo/blob/main/109_Selfie_Segmentation/selfie-model-card.pdf) says so, but this might simply not be accurate,
it also states, there would be 2 output layers, but i can find only one
also, if you look at the IE xml files in the zip, those clearly have BCHW order

i checked your code using the `model_float32.pb` and got decent results without reordering the blob channels

however, i could NOT import the `model_float32.onnx` in 4.5.5-dev:
```
[ERROR:0@0.693] global C:\p\opencv\modules\dnn\src\onnx\onnx_importer.cpp (909) handleNode DNN/ONNX: ERROR during processing node with 3 inputs and 1 outputs: [Clip]:(Relu6:0) from domain='ai.onnx'
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.5.5-dev) Error: Unspecified error (> Node [Clip@ai.onnx]:(Relu6:0) parse error: OpenCV(4.5.5-dev) C:\p\opencv\modules\dnn\src\onnx\onnx_importer.cpp:1613: error: (-2:Unspecified error) in function 'void cv::dnn::dnn4_v20211220::ONNXImporter::parseClip(cv::dnn::dnn4_v20211220::LayerParams&, const opencv_onnx::NodeProto&)'
> >  (expected: 'node_proto.input_size() == 1'), where
> >     'node_proto.input_size()' is 3
> > must be equal to
> >     '1' is 1
> ) in handleNode, file C:\p\opencv\modules\dnn\src\onnx\onnx_importer.cpp, line 928
```
while this works ok with a previous one (`pip install opencv-python==4.5.4.60`):

```
>>> import cv2
>>> print(cv2.__version__)
4.5.4
>>> cv2.dnn.readNet("model_float32.onnx")
<dnn_Net 0x7f99b27e99b0>
```

there seems to be some regression here

@cesarpgouveia , can you check the version again ? sure it's 4.5.5 ?

--- a/modules/dnn/src/onnx/onnx_importer.cpp
+++ b/modules/dnn/src/onnx/onnx_importer.cpp
@@ -15,6 +15,9 @@
 #define CV_LOG_STRIP_LEVEL CV_LOG_LEVEL_VERBOSE + 1
 #include <opencv2/core/utils/logger.hpp>
 
+#include <opencv2/core/utils/configuration.private.hpp>
+
+
 #ifdef HAVE_PROTOBUF
 
 #include <iostream>
@@ -78,6 +81,7 @@ public:
     ONNXImporter(Net& net, const char *onnxFile)
         : dstNet(net), dispatch(buildDispatchMap())
         , onnx_opset(0)
+        , useLegacyNames(getParamUseLegacyNames())
     {
         hasDynamicShapes = false;
         CV_Assert(onnxFile);
@@ -100,6 +104,7 @@ public:
     ONNXImporter(Net& net, const char* buffer, size_t sizeBuffer)
         : dstNet(net), dispatch(buildDispatchMap())
         , onnx_opset(0)
+        , useLegacyNames(getParamUseLegacyNames())
     {
         hasDynamicShapes = false;
         CV_LOG_DEBUG(NULL, "DNN/ONNX: processing in-memory ONNX model (" << sizeBuffer << " bytes)");
@@ -196,7 +201,17 @@ private:
 
     int onnx_opset;  // OperatorSetIdProto for 'onnx' domain
     void parseOperatorSet();
+
+
+    bool useLegacyNames;
+    bool getParamUseLegacyNames()
+    {
+        bool param = utils::getConfigurationParameterBool("OPENCV_DNN_ONNX_USE_LEGACY_NAMES", false);
+        return param;
+    }
+    const std::string extractNodeName(const opencv_onnx::NodeProto& node_proto);
 };
+
 
 inline void replaceLayerParam(LayerParams& layerParams, const String& oldKey, const String& newKey)
 {
@@ -720,12 +735,13 @@ void ONNXImporter::populateNet()
     CV_LOG_DEBUG(NULL, "DNN/ONNX: import completed!");
 }
 
-static
-const std::string extractNodeName(const opencv_onnx::NodeProto& node_proto)
+const std::string ONNXImporter::extractNodeName(const opencv_onnx::NodeProto& node_proto)
 {
     // We need to rework DNN outputs API, this is a workaround for #21698
     if (node_proto.has_name() && !node_proto.name().empty())
     {
+        if (useLegacyNames)
+            return node_proto.name();
         return cv::format("onnx_node!%s", node_proto.name().c_str());
     }
     for (int i = 0; i < node_proto.output_size(); ++i)
@@ -736,6 +752,8 @@ const std::string extractNodeName(const opencv_onnx::NodeProto& node_proto)
         // the second method is to use an empty string in place of an input or output name.
         if (!name.empty())
         {
+            if (useLegacyNames)
+                return name.c_str();
             return cv::format("onnx_node_output_%d!%s", i, name.c_str());
         }
     }

--- a/modules/dnn/src/onnx/onnx_importer.cpp
+++ b/modules/dnn/src/onnx/onnx_importer.cpp
@@ -721,11 +721,12 @@ void ONNXImporter::populateNet()
 }
 
 static
-const std::string& extractNodeName(const opencv_onnx::NodeProto& node_proto)
+const std::string extractNodeName(const opencv_onnx::NodeProto& node_proto)
 {
+    // We need to rework DNN outputs API, this is a workaround for #21698
     if (node_proto.has_name() && !node_proto.name().empty())
     {
-        return node_proto.name();
+        return cv::format("onnx_node!%s", node_proto.name().c_str());
     }
     for (int i = 0; i < node_proto.output_size(); ++i)
     {
@@ -735,7 +736,7 @@ const std::string& extractNodeName(const opencv_onnx::NodeProto& node_proto)
         // the second method is to use an empty string in place of an input or output name.
         if (!name.empty())
         {
-            return name;
+            return cv::format("onnx_node_output_%d!%s", i, name.c_str());
         }
     }
     CV_Error(Error::StsAssert, "Couldn't deduce Node name.");

--- a/modules/dnn/test/test_onnx_importer.cpp
+++ b/modules/dnn/test/test_onnx_importer.cpp
@@ -1389,6 +1389,11 @@ TEST_P(Test_ONNX_layers, DivConst)
     testONNXModels("div_const");
 }
 
+TEST_P(Test_ONNX_layers, OutputRegistration)
+{
+    testONNXModels("output_registration", npy, 0, 0, false, true, 2);
+}
+
 INSTANTIATE_TEST_CASE_P(/*nothing*/, Test_ONNX_layers, dnnBackendsAndTargets());
 
 class Test_ONNX_nets : public Test_ONNX_layers

